src/reporter.ts

```diff
@@ -0,0 +1,251 @@
+      this.api = new SauceLabs({
+        user: process.env.SAUCE_USERNAME,
+        key: process.env.SAUCE_ACCESS_KEY,
+        region: this.region,
+        headers: {
+          'User-Agent': `cucumber-reporter/${reporterVersion}`
+        },
+      });
+      this.api.tld = this.region === 'staging' ? 'net' : 'com';
+    }
+    if (process.env.SAUCE_VIDEO_START_TIME) {
+      this.videoStartTime = new Date(process.env.SAUCE_VIDEO_START_TIME).getTime();
+    }
+
+    config.eventBroadcaster?.on('envelope', (envelope: Envelope) => {
+      if (envelope.testCaseFinished) {
+        this.logTestCase(envelope.testCaseFinished)
+      }
+      if (envelope.testRunFinished) {
+        this.endedAt = new Date().toISOString();
+        this.logTestRun(envelope.testRunFinished)
+      }
+    })
+  }
+
+  logTestCase(testCaseFinished: { testCaseStartedId: any }) {
+    const testCaseAttempt = this.eventDataCollector.getTestCaseAttempt(testCaseFinished.testCaseStartedId)
+    const parsed = formatterHelpers.parseTestCaseAttempt({
+      //cwd: this.cwd,
+      snippetBuilder: this.snippetBuilder, 
+      supportCodeLibrary: this.supportCodeLibrary,
+      testCaseAttempt 
+    })
+    const suite = new Suite(parsed.testCase?.sourceLocation?.uri || '')
+    const curr = new Suite(parsed.testCase?.name);
+    curr.metadata = {
+      attempt: parsed.testCase?.attempt,
+      sourceLocation: parsed.testCase?.sourceLocation,
+    }
+    this.consoleLog.push(`${curr.name}\t#${suite.name}`)
+
+    parsed.testSteps.forEach((testStep) => {
+      this.consoleLog.push('  ' + testStep.keyword + (testStep.text || '') + ' - ' + Status[testStep.result.status]);
+      const test = new Test(`${testStep.keyword}${testStep.text || ''}`)
+      const testStatus = testStep.result?.status?.toLowerCase();
+      test.status = testStatus === 'skipped' ? SauceStatus.Skipped : (testStatus === 'passed' ? SauceStatus.Passed : SauceStatus.Failed);
+      test.output = testStep.result?.message;
+      test.duration = this.durationToMilliseconds(testStep.result?.duration);
+      test.attachments = [];
+      testStep.attachments.forEach((attachment: any) => {
+        this.assets.push({
+          filename: attachment.testCaseStartedId + '.log',
+          data: Buffer.from(attachment.body),
+        })
+        test.attachments?.push({
+          name: attachment.testCaseStartedId + '.log',
+          contentType: attachment.mediaType,
+          path: ''
+        })
+      })
+      curr.addTest(test);
+      curr.status = this.testRun.computeStatus()
+    })
+    suite.addSuite(curr)
+    this.testRun.addSuite(suite);
+  }
+
+  async logTestRun(testRunFinished: { success: any; }) {
+    this.passed = testRunFinished.success;
+    this.reportToFile(this.testRun)
+
+    const id = await this.reportToSauce();
+    this.consoleLog.push(testRunFinished.success ? 'SUCCESS' : 'FAILURE')
+    this.logSauceJob(id as string)
```

Comment:
imo, when you do this kind of casting, you lose some of the helpfulness of typescript. `id` has a type of `string | undefined`. By doing `id as string` you've made the compiler happy, but not in a way that makes the code more reliable, i.e. ignoring `undefined` in js is generally a potential bug. I see you do a nullish check in `logSauceJob` so the logic of your implementation is correct. But this cast is a code smell to me. As is the fact that the method signature of `logSauceJob` says the argument is not optional, but you do a null check anyways. These kinds of inconsistencies over time make the code harder to understand and harder to correctly refactor.

Reply:
Gotcha. Make sense! Will fix that.